src/com/palmergames/bukkit/towny/tasks/HourlyTimerTask.java

```diff
@@ -56,5 +60,25 @@ private void decrementJailedHours() {
 					resident.setJailHours(resident.getJailHours() - 1);
 					resident.save();
 				}
+		if (TownyEconomyHandler.isActive() && TownySettings.hourlyJailFee() > 0)
+		{
+			// Check if towns can afford upkeep for prisoners and if not to release them
+			for (Resident resident : new ArrayList<>(universe.getJailedResidentMap()))
+			{
+				Town town = resident.getJailTown();
+				double hourlyJailFee = TownySettings.hourlyJailFee();
+				if (!town.getAccount().canPayFromHoldings(hourlyJailFee))
+				{
+					Bukkit.getScheduler().runTaskLater(plugin, () -> JailUtil.unJailResident(resident, UnJailReason.INSUFFICIENT_FUNDS), 20);
+				}
+				else
+				{
+					TownyMessaging.sendPrefixedTownMessage(town, Translatable.of("msg_x_has_been_withdrawn_for_upkeep_of_prisoner_x", hourlyJailFee,resident));
+					String reason = "Jailee Upkeep for " + resident.getName();
+					town.getAccount().withdraw(hourlyJailFee, reason);
+				}
+					
+			}
+		}
```

Comment:
An issue with this code is that it is running outside of the for loop run above it. Any resident that has been unjailed is not going to be unjailed properly until 20 ticks later.

I think the work around is to use one for loop, that checks if they have no jail time left (unjailing if needed,) then checks if there is a jail fee and if the town can pay it (unjailing if needed,) and if they are still jailed, decrement their jail hours.

Reply:
Method has been redesigned to all run within the same for loop.